src/kessel/auth/auth.py

```diff
@@ -0,0 +1,154 @@
+
+    Returns:
+        OIDCDiscoveryMetadata containing the discovered endpoints.
+
+    Raises:
+        requests.exceptions.RequestException: If the discovery document cannot be retrieved.
+        ValueError: If the response is not valid JSON.
+    """
+    discovery_url = f"{issuer_url.rstrip('/')}/.well-known/openid-configuration"
+    response = requests.get(discovery_url, timeout=10)
+    response.raise_for_status()
+    config = response.json()
+
+    return OIDCDiscoveryMetadata(config)
+
+
+class OAuth2ClientCredentials:
+    """
+    OAuth2ClientCredentials class for handling the OAuth 2.0 Client Credentials flow.
+
+    Integrates with the google-auth and requests-oauthlib library to fetch an access token
+    from a specified token endpoint with automatic refreshing.
+
+    This class only accepts a direct token URL. For OIDC discovery, use the
+    fetch_oidc_discovery function to obtain the token endpoint first.
+    """
+
+    def __init__(
+        self,
+        client_id: str,
+        client_secret: str,
+        token_endpoint: str,
+    ):
+        """
+        Initializes the OAuth2ClientCredentials.
+
+        Args:
+            client_id: The client ID for the application.
+            client_secret: The client secret for the application.
+            token_endpoint: The direct token endpoint URL.
+        """
+        self._token_endpoint = token_endpoint
+        self._client_id = client_id
+        self._client_secret = client_secret
+
+        client = BackendApplicationClient(client_id=self._client_id)
+        self._session = OAuth2Session(client=client)
+
+        self.token = None
+        self.expiry = None
+
+    def get_token(self, force_refresh: bool = False) -> Tuple[str, int]:
+        """
+        Get a valid access token, refreshing if necessary or forced.
+
+        Args:
+            force_refresh: If True, forces token refresh regardless of expiry.
+
+        Returns:
+            Tuple containing (access_token, expires_in).
+        """
+        current_time = datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None)
+
+        if (
+            force_refresh
+            or self.token is None
+            or self.expiry is None
+            or self.expiry <= current_time + datetime.timedelta(seconds=300)
+        ):
+            # Refresh the token
+            token_data = self._session.fetch_token(
+                token_url=self._token_endpoint,
+                client_id=self._client_id,
+                client_secret=self._client_secret,
+            )
+
+            self.token = token_data.get("access_token")
+            expires_in = token_data.get("expires_in", 0)
+            self.expiry = current_time + datetime.timedelta(seconds=expires_in)
+
+        remaining_seconds = int(
+            (
+                self.expiry - datetime.datetime.now(datetime.timezone.utc).replace(tzinfo=None)
+            ).total_seconds()
+        )
+        return (self.token, remaining_seconds)
```

Comment:
Could you use `expires_in` instead of calculating `remaining_seconds`?

```suggestion
        return (self.token, expires_in)
```

Reply:
We could update the spec and return the "expires_at" instead - I also have to do the same (calculate expires_in)

expires_in is what the token return, but expires_at is more convenient and is already local to our time.

Reply:
`expires_at` would be much simpler for this setup, already have the datetime object of expected expiration required for google-auth.